### ISSUE TITLE
Memory leak in tests detected by Address Sanitizer

### ISSUE THREAD

**Issue by Klaim** [openSUSE/libsolv#496]
# Context 

As part of investigation of some issues in [libmamba](https://github.com/mamba-org/mamba) which depends on `libsolv` I used Address Sanitizer (ASAN) and spotted a memory leak report coming from the lib. While there is probably mishandling on `libmamba`'s side, to be sure I attempted to build `libsolv` with ASAN and found a few tests leaking.

The following is on Ubuntu, using conda/mamba as environment, notably `gcc==10.3`

# How To Reproduce

0. clone `libsolve`, create a build directory
1. cmake invocation I used:
    ```
    cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_CXX_FLAGS_DEBUG="-g -O0 -fno-omit-frame-pointer -fsanitize=address" -DLINK_FLAGS="-fno-omit-frame-pointer -fsanitize=address" -DBUILD_TESTING=ON -DBUILD_SHARED=ON -S ../ -B ./
    ```
 2. Build and run the tests: `make && make test`
  
# Observed
 
The output:
```
[...]
96% tests passed, 1 tests failed out of 26

Total Test time (real) =   4.51 sec

The following tests FAILED:
          3 - choicerules (Failed)
Errors while running CTest
Output from these tests are in: /home/joel/work/libs/libsolv/build-debug/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
```
In the log file we can find:
```
3/26 Test: choicerules
Command: "/home/joel/work/libs/libsolv/test/runtestcases.sh" "/home/joel/work/libs/libsolv/build-debug/tools/testsolv" "/home/joel/work/libs/libsolv/test/testcases/choicerules"
Directory: /home/joel/work/libs/libsolv/build-debug/test
"choicerules" start time: Apr 27 16:40 CEST
Output:
----------------------------------------------------------
testcase_read: missing feature 'complex_deps'
choice1b.t .......................................   Skipped

=================================================================
==255147==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 1024 byte(s) in 1 object(s) allocated from:
    #0 0x7fdf5d685ea8 in __interceptor_malloc ../../../../libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x7fdf5d5884f2 in solv_malloc /home/joel/work/libs/libsolv/src/util.c:38
    #2 0x7fdf5d5edf9a in read_inline_file /home/joel/work/libs/libsolv/ext/testcase.c:2005
    #3 0x7fdf5d5ef8b4 in testcase_read /home/joel/work/libs/libsolv/ext/testcase.c:2437
    #4 0x56401e62b783 in main /home/joel/work/libs/libsolv/tools/testsolv.c:159
    #5 0x7fdf5d32ffcf in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58

SUMMARY: AddressSanitizer: 1024 byte(s) leaked in 1 allocation(s).
choice1.t ........................................***Failed
choice2b.t .......................................   Passed
choice2.t ........................................   Passed
choice3b.t .......................................   Passed
choice3.t ........................................   Passed
choice4.t ........................................   Passed

=================================================================
==255159==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 1 byte(s) in 1 object(s) allocated from:
    #0 0x7f8ffc3c6fb9 in __interceptor_calloc ../../../../libsanitizer/asan/asan_malloc_linux.cpp:154
    #1 0x7f8ffc2c968e in solv_calloc /home/joel/work/libs/libsolv/src/util.c:79
    #2 0x7f8ffc289110 in map_init /home/joel/work/libs/libsolv/src/bitmap.c:24
    #3 0x7f8ffc2e6d6c in solver_disablechoicerules /home/joel/work/libs/libsolv/src/rules.c:3577
    #4 0x7f8ffc294b0d in solver_check_unneeded_choicerules /home/joel/work/libs/libsolv/src/solver.c:2651
    #5 0x7f8ffc2953fc in solver_run_sat /home/joel/work/libs/libsolv/src/solver.c:2845
    #6 0x7f8ffc29bc1c in solver_solve /home/joel/work/libs/libsolv/src/solver.c:4267
    #7 0x56169069ac05 in main /home/joel/work/libs/libsolv/tools/testsolv.c:241
    #8 0x7f8ffc070fcf in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58

SUMMARY: AddressSanitizer: 1 byte(s) leaked in 1 allocation(s).
choice5.t ........................................***Failed
```

[LastTest.log](https://github.com/openSUSE/libsolv/files/8573133/LastTest.log)
[LastTestsFailed.log](https://github.com/openSUSE/libsolv/files/8573134/LastTestsFailed.log)

Note that the tests pass without ASAN reports, as the test checks the ouput of the program.

# Expected

No failing test and no report coming from ASAN.

**Comment by banjiuqingshan** [comment]
I met the same problem. Do you know the reason now

**Comment by Klaim** [comment]
I didn't have time to take a deep dive into libsolv's code yet, but the report suggests that a test is leaking after using libsolv's allocation function, and the map implementation is leaking too, but that should be on the map usage side I guess.

**Comment by mlschroe** [comment]
This has been fixed with commit eb94d1c0c750dcdee0e82a8b649c4eae22edd50f, so I'm closing this issue.
